Not possible to create more than 3 network interfaces!

Hi there,

### Terraform Version
Terraform v0.11.1

### vSphere Provider Version
+ provider.vsphere v1.0.1
+ provider.vsphere v1.0.2

### Affected Resource(s)

Please list the resources as a list, for example:
- `vsphere_virtual_machine`

### Terraform Configuration Files

```hcl
data "vsphere_datacenter" "dc" {
  name = "ha-datacenter"
}

data "vsphere_datastore" "datastore" {
  name          = "datastore1"
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

data "vsphere_resource_pool" "pool" {}

data "vsphere_network" "mgmt_lan" {
  name          = "MGMT LAN"
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

data "vsphere_network" "red" {
  name          = "Behind server"
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

resource "vsphere_virtual_machine" "test2" {
  name             = "test2"
  resource_pool_id = "${data.vsphere_resource_pool.pool.id}"
  datastore_id     = "${data.vsphere_datastore.datastore.id}"

  num_cpus   = 1
  memory     = 2048
  wait_for_guest_net_timeout = 0
  guest_id = "centos7_64Guest"
  nested_hv_enabled =true
  network_interface {
   network_id     = "${data.vsphere_network.mgmt_lan.id}"
   adapter_type   = "vmxnet3"
   use_static_mac = true
   mac_address    = "00:50:56:17:01:01"
  }

  network_interface {
   network_id     = "${data.vsphere_network.mgmt_lan.id}"
   adapter_type   = "vmxnet3"
   use_static_mac = true
   mac_address    = "00:50:56:17:02:01"
  }

  network_interface {
   network_id     = "${data.vsphere_network.mgmt_lan.id}"
   adapter_type   = "vmxnet3"
   use_static_mac = true
   mac_address    = "00:50:56:17:03:01"
  }

  network_interface {
   network_id     = "${data.vsphere_network.red.id}"
   adapter_type   = "vmxnet3"
   use_static_mac = true
   mac_address    = "00:50:56:17:05:01"
  }

  disk {
   size             = 16
   name             = "test2.vmdk"
   eagerly_scrub    = false
   thin_provisioned = true
  }
}

```

### Debug Output


### Panic Output
panic: runtime error: index out of range
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4:
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: goroutine 111 [running]:
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: github.com/terraform-providers/terraform-provider-vsphere/vsphere/internal/virtualdevice.(*NetworkInterfaceSubresource).assignEthernetCard(0xc42000c560, 0xc42013a600, 0xe, 0x10, 0x2038340, 0xc4203655e0, 0x2037fc0, 0xc420242070, 0xc4203655e0, 0x0)
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4:       /opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go:871 +0x28e
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: github.com/terraform-providers/terraform-provider-vsphere/vsphere/internal/virtualdevice.(*NetworkInterfaceSubresource).Create(0xc42000c560, 0xc42013a600, 0xe, 0x10, 0x0, 0x0, 0x0, 0x0, 0x0)
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4:       /opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go:587 +0x406
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: github.com/terraform-providers/terraform-provider-vsphere/vsphere/internal/virtualdevice.NetworkInterfaceApplyOperation(0xc420520690, 0xc420593750, 0xc42013a600, 0xe, 0x10, 0x1, 0x1, 0x1, 0xc420581380, 0x1, ...)
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4:       /opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go:201 +0x1339
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: github.com/terraform-providers/terraform-provider-vsphere/vsphere.applyVirtualDevices(0xc420520690, 0xc420593750, 0xc42013a600, 0xb, 0x10, 0x9, 0x10, 0x0, 0x0, 0x0)
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4:       /opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vsphere/resource_vsphere_virtual_machine.go:887 +0x32d
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: github.com/terraform-providers/terraform-provider-vsphere/vsphere.resourceVSphereVirtualMachineCreateBare(0xc420520690, 0x11f23e0, 0xc420592d00, 0x1162800, 0xc420342ca0, 0x0)
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4:       /opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vsphere/resource_vsphere_virtual_machine.go:630 +0xa56
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: github.com/terraform-providers/terraform-provider-vsphere/vsphere.resourceVSphereVirtualMachineCreate(0xc420520690, 0x11f23e0, 0xc420592d00, 0x24, 0x20d4e60)
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4:       /opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vsphere/resource_vsphere_virtual_machine.go:212 +0x40a
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/helper/schema.(*Resource).Apply(0xc42057dce0, 0xc420059b30, 0xc4203ff6c0, 0x11f23e0, 0xc420592d00, 0x1, 0xc4202db6e0, 0xc4204014a0)
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4:       /opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/helper/schema/resource.go:227 +0x364
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/helper/schema.(*Provider).Apply(0xc4205e0310, 0xc420059a90, 0xc420059b30, 0xc4203ff6c0, 0x7f1efb225000, 0x0, 0x18)
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4:       /opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/helper/schema/provider.go:283 +0xa4
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/plugin.(*ResourceProviderServer).Apply(0xc420517860, 0xc4203fe660, 0xc4202df500, 0x0, 0x0)
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4:       /opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform/plugin/resource_provider.go:527 +0x57
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: reflect.Value.call(0xc4203b6240, 0xc4204ac000, 0x13, 0x1477d1e, 0x4, 0xc420041f20, 0x3, 0x3, 0x0, 0x0, ...)
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4:       /usr/local/go/src/reflect/value.go:434 +0x906
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: reflect.Value.Call(0xc4203b6240, 0xc4204ac000, 0x13, 0xc420022720, 0x3, 0x3, 0xc4203b7560, 0x0, 0x0)
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4:       /usr/local/go/src/reflect/value.go:302 +0xa4
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: net/rpc.(*service).call(0xc4205a0100, 0xc420058050, 0xc42053a070, 0xc420458000, 0xc420516580, 0x1110100, 0xc4203fe660, 0x16, 0x1110140, 0xc4202df500, ...)
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4:       /usr/local/go/src/net/rpc/server.go:381 +0x142
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4: created by net/rpc.(*Server).ServeCodec
2017-12-06T14:37:57.290+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.0.1_x4:       /usr/local/go/src/net/rpc/server.go:475 +0x36b
2017/12/06 14:37:57 [TRACE] root: eval: *terraform.EvalWriteState
2017/12/06 14:37:57 [TRACE] root: eval: *terraform.EvalApplyProvisioners
2017/12/06 14:37:57 [TRACE] root: eval: *terraform.EvalIf
2017/12/06 14:37:57 [TRACE] root: eval: *terraform.EvalWriteState
2017/12/06 14:37:57 [TRACE] root: eval: *terraform.EvalWriteDiff
2017/12/06 14:37:57 [TRACE] root: eval: *terraform.EvalApplyPost
2017/12/06 14:37:57 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* vsphere_virtual_machine.test2: unexpected EOF

### Expected Behavior

Create a new VM with 4 nic

### Actual Behavior
Only possible to create a VM with 3 nic's!

Not even possible is you create a VM with 3 nic, **then add 1 more**, **crash**

### Steps to Reproduce
1. `terraform apply`

### Important Factoids
verified on both esxi 5.5 and 6.5, same behavior.

### References

--- a/vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go
+++ b/vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go
@@ -870,7 +870,7 @@ func (r *NetworkInterfaceSubresource) assignEthernetCard(l object.VirtualDeviceL
 
 	// Now that we know which units are used, we can pick one
 	newUnit := int32(r.Index) + pciDeviceOffset
-	if units[newUnit] {
+	if units[newUnit-pciDeviceOffset] {
 		return fmt.Errorf("device unit at %d is currently in use on the PCI bus", newUnit)
 	}
 

--- a/vsphere/resource_vsphere_virtual_machine_test.go
+++ b/vsphere/resource_vsphere_virtual_machine_test.go
@@ -281,6 +281,26 @@ func TestAccResourceVSphereVirtualMachine(t *testing.T) {
 			},
 		},
 		{
+			"maximum number of nics",
+			resource.TestCase{
+				PreCheck: func() {
+					testAccPreCheck(tp)
+					testAccResourceVSphereVirtualMachinePreCheck(tp)
+				},
+				Providers:    testAccProviders,
+				CheckDestroy: testAccResourceVSphereVirtualMachineCheckExists(false),
+				Steps: []resource.TestStep{
+					{
+						Config: testAccResourceVSphereVirtualMachineConfigMaxNIC(),
+						Check: resource.ComposeTestCheckFunc(
+							testAccResourceVSphereVirtualMachineCheckExists(true),
+							testAccResourceVSphereVirtualMachineCheckNICCount(10),
+						),
+					},
+				},
+			},
+		},
+		{
 			"upgrade cpu and ram",
 			resource.TestCase{
 				PreCheck: func() {
@@ -1808,6 +1828,28 @@ func testAccResourceVSphereVirtualMachineCheckVmdkDatastore(name, expected strin
 	}
 }
 
+// testAccResourceVSphereVirtualMachineCheckNICCount checks the number of NICs
+// on the virtual machine.
+func testAccResourceVSphereVirtualMachineCheckNICCount(expected int) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		props, err := testGetVirtualMachineProperties(s, "vm")
+		if err != nil {
+			return err
+		}
+
+		var actual int
+		for _, dev := range props.Config.Hardware.Device {
+			if _, ok := dev.(types.BaseVirtualEthernetCard); ok {
+				actual++
+			}
+		}
+		if expected != actual {
+			return fmt.Errorf("expected %d number of NICs, got %d", expected, actual)
+		}
+		return nil
+	}
+}
+
 func testAccResourceVSphereVirtualMachineConfigBasic() string {
 	return fmt.Sprintf(`
 variable "datacenter" {
@@ -2373,6 +2415,105 @@ resource "vsphere_virtual_machine" "vm" {
   num_cpus = 4
   memory   = 8192
   guest_id = "other3xLinux64Guest"
+
+  network_interface {
+    network_id = "${data.vsphere_network.network.id}"
+  }
+
+  disk {
+    name = "terraform-test.vmdk"
+    size = 20
+  }
+}
+`,
+		os.Getenv("VSPHERE_DATACENTER"),
+		os.Getenv("VSPHERE_RESOURCE_POOL"),
+		os.Getenv("VSPHERE_NETWORK_LABEL_PXE"),
+		os.Getenv("VSPHERE_DATASTORE"),
+	)
+}
+
+func testAccResourceVSphereVirtualMachineConfigMaxNIC() string {
+	return fmt.Sprintf(`
+variable "datacenter" {
+  default = "%s"
+}
+
+variable "resource_pool" {
+  default = "%s"
+}
+
+variable "network_label" {
+  default = "%s"
+}
+
+variable "datastore" {
+  default = "%s"
+}
+
+data "vsphere_datacenter" "dc" {
+  name = "${var.datacenter}"
+}
+
+data "vsphere_datastore" "datastore" {
+  name          = "${var.datastore}"
+  datacenter_id = "${data.vsphere_datacenter.dc.id}"
+}
+
+data "vsphere_resource_pool" "pool" {
+  name          = "${var.resource_pool}"
+  datacenter_id = "${data.vsphere_datacenter.dc.id}"
+}
+
+data "vsphere_network" "network" {
+  name          = "${var.network_label}"
+  datacenter_id = "${data.vsphere_datacenter.dc.id}"
+}
+
+resource "vsphere_virtual_machine" "vm" {
+  name             = "terraform-test"
+  resource_pool_id = "${data.vsphere_resource_pool.pool.id}"
+  datastore_id     = "${data.vsphere_datastore.datastore.id}"
+
+  num_cpus = 2
+  memory   = 2048
+  guest_id = "other3xLinux64Guest"
+
+  network_interface {
+    network_id = "${data.vsphere_network.network.id}"
+  }
+
+  network_interface {
+    network_id = "${data.vsphere_network.network.id}"
+  }
+
+  network_interface {
+    network_id = "${data.vsphere_network.network.id}"
+  }
+
+  network_interface {
+    network_id = "${data.vsphere_network.network.id}"
+  }
+
+  network_interface {
+    network_id = "${data.vsphere_network.network.id}"
+  }
+
+  network_interface {
+    network_id = "${data.vsphere_network.network.id}"
+  }
+
+  network_interface {
+    network_id = "${data.vsphere_network.network.id}"
+  }
+
+  network_interface {
+    network_id = "${data.vsphere_network.network.id}"
+  }
+
+  network_interface {
+    network_id = "${data.vsphere_network.network.id}"
+  }
 
   network_interface {
     network_id = "${data.vsphere_network.network.id}"
@@ -3295,6 +3436,7 @@ resource "vsphere_virtual_machine" "vm" {
 		os.Getenv("VSPHERE_USE_LINKED_CLONE"),
 	)
 }
+
 func testAccResourceVSphereVirtualMachineConfigCloneTimeZone(zone string) string {
 	return fmt.Sprintf(`
 variable "datacenter" {

